### PR TITLE
[GPU] Add IE Core GPU plugin metric to query overall memory statistics for GPU device

### PR DESCRIPTION
**Pull request by andrew-k-park**
### Details:
 - Add IE Core GPU plugin metric to query overall memory statistics for GPU device
 - Follow up of comments for https://github.com/openvinotoolkit/openvino/pull/7758

### Tickets:
 - 56842

### CODE REVIEW

File: docs/snippets/InferenceEngine_QueryAPI7.cpp [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#include <ie_core.hpp>
+
+int main() {
+using namespace InferenceEngine;
+//! [part7]
+InferenceEngine::Core core;
+auto network = core.ReadNetwork("sample.xml");
+auto exeNetwork = core.LoadNetwork(network, "GPU");
+std::map<std::string, uint64_t> statistics_map = core.GetMetric("GPU", GPU_METRIC_KEY(MEMORY_STATISTICS));
```
Comment (yeonbok) [inline docs/snippets/InferenceEngine_QueryAPI7.cpp:9]: BTW I am curious that is this really added to InferenceEngine_Query ? 
Seems that currently MEMORY_STATISTICS is only available as **GPU_**METRIC_KEY.

Comment (yeonbok) [reply docs/snippets/InferenceEngine_QueryAPI7.cpp:9]: @vladimir-paramuzov  How do you think the location of this metric?
I am putting the GPU specific metric to GPU.md, as follows:
https://github.com/openvinotoolkit/openvino/pull/8226/files

Comment (vladimir-paramuzov) [reply docs/snippets/InferenceEngine_QueryAPI7.cpp:9]: @yeonbok I like the approach with GPU specific metrics in GPU.md

File: inference-engine/src/cldnn_engine/cldnn_engine.cpp [hunk 2]
```diff
@@ -258,7 +277,9 @@ IExecutableNetworkInternal::Ptr clDNNEngine::LoadExeNetworkImpl(const InferenceE
     auto transformedNetwork = CloneAndTransformNetwork(network, conf);
     {
         OV_ITT_SCOPED_TASK(itt::domains::CLDNNPlugin, "clDNNEngine::LoadExeNetworkImpl::CreateExeNetwork");
-        return std::make_shared<CLDNNExecNetwork>(transformedNetwork, context, conf);
+        CLDNNExecNetwork::Ptr exeNetwork = std::make_shared<CLDNNExecNetwork>(transformedNetwork, context, conf);
+        UpdateStatistics(context);
```
Comment (vladimir-paramuzov) [inline inference-engine/src/cldnn_engine/cldnn_engine.cpp:281]: What about multiple calls of GetMetric for executable network? Does it still accumulate memory consumption for all networks or it returns per-network value?

File: docs/snippets/InferenceEngine_QueryAPI7.cpp (hunk 1 above)
Comment (andrew-k-park) [reply docs/snippets/InferenceEngine_QueryAPI7.cpp:9]: @yeonbok @vladimir-paramuzov Let me move this update for document and code snippet to GPU.md

File: inference-engine/src/cldnn_engine/cldnn_engine.cpp (hunk 2 above)
Comment (andrew-k-park) [reply inference-engine/src/cldnn_engine/cldnn_engine.cpp:281]: It still accumulates memory usages for networks allocated through the same context. So way to separate the memory usage of networks sharing the same context is currently under design and PoC, and the purpose of this PR is to query the overall memory usage statistics of all networks allocated through different contexts or the same context. BTW what does "multiple calls of GetMetric for executable network" mean? Could you please elaborate?

File: docs/snippets/InferenceEngine_QueryAPI7.cpp (hunk 1 above)
Comment (andrew-k-park) [reply docs/snippets/InferenceEngine_QueryAPI7.cpp]: done

File: inference-engine/tests/functional/plugin/gpu/shared_tests_instances/behavior/plugin/core_integration.cpp [hunk 3]
```diff
@@ -153,6 +153,30 @@ INSTANTIATE_TEST_SUITE_P(
...
+using IEClassGetMetricTest_GPU_MEMORY_STATISTICS = BehaviorTestsUtils::IEClassBaseTestP;
+TEST_P(IEClassGetMetricTest_GPU_MEMORY_STATISTICS, GetMetricAndPrintNoThrow) {
+    SKIP_IF_CURRENT_TEST_IS_DISABLED()
+    InferenceEngine::Core ie;
+    InferenceEngine::Parameter p;
+
+    InferenceEngine::ExecutableNetwork exeNetwork = ie.LoadNetwork(simpleCnnNetwork, deviceName);
+
+    ASSERT_NO_THROW(p = ie.GetMetric(deviceName, GPU_METRIC_KEY(MEMORY_STATISTICS)));
+    std::map<std::string, uint64_t> t = p;
```
Comment (vladimir-paramuzov) [inline inference-engine/tests/functional/plugin/gpu/shared_tests_instances/behavior/plugin/core_integration.cpp:165]: It'd be great to have more functional tests and some result value checks.
E.g.:
1. create core -> get stat -> check that it's 0 -> load network -> check that it's != 0 -> free network -> check that it's zero again.
2. load network -> get stat -> load second net instance -> check that mem stat doubles
etc

Comment (andrew-k-park) [reply inference-engine/tests/functional/plugin/gpu/shared_tests_instances/behavior/plugin/core_integration.cpp:165]: @vladimir-paramuzov I have simple question regarding how to free network for increasing test coverage. Is there a way or sample to manually free exeNetwork created through LoadNetwork before it is automatically released? Furthermore, the shared pointer of exeNetwork, impl_ was found to be inaccessible.

Comment (andrew-k-park) [reply inference-engine/tests/functional/plugin/gpu/shared_tests_instances/behavior/plugin/core_integration.cpp:165]: @vladimir-paramuzov ping again. :-)

Comment (vladimir-paramuzov) [reply inference-engine/tests/functional/plugin/gpu/shared_tests_instances/behavior/plugin/core_integration.cpp:165]: Option 1:
```cpp
    // check memory statistics
    {
        auto exec_net1 = ie.LoadNetwork(net1, CommonTestUtils::DEVICE_GPU);
        {
             auto exec_net2 = ie.LoadNetwork(net2, CommonTestUtils::DEVICE_GPU);
            // check memory statistics
        }
        // check memory statistics
    }
    // check memory statistics
```
Option 2:
```cpp
    // check memory statistics
    std::shared_ptr<ExecutableNetwork> exec_net1 = nullptr;

    {
        auto exec_net_tmp = ie.LoadNetwork(net, CommonTestUtils::DEVICE_GPU);
        exec_net1.reset(new ExecutableNetwork(exec_net_tmp));
    }
    // check memory statistics
    exec_net1.reset(nullptr);
    // check memory statistics 
```

Comment (andrew-k-park) [reply inference-engine/tests/functional/plugin/gpu/shared_tests_instances/behavior/plugin/core_integration.cpp:165]: Added test cases using the first option.

**Comment by yeonbok** [comment]
@andrew-k-park  as we discussed in the offline meeting, we'll disble the execnetwork's metric in this PR. That metric will be added once the final impl is done.

**Comment by andrew-k-park** [comment]
> @andrew-k-park as we discussed in the offline meeting, we'll disable the ExecNetwork's metric in this PR. That metric will be added once the final impl is done. So please disable ExecNet metric in this PR.

**Comment by andrew-k-park** [comment]
> @andrew-k-park as we discussed in the offline meeting, we'll disable the ExecNetwork's metric in this PR. That metric will be added once the final impl is done. So please disable ExecNet metric in this PR.

Agree. As comment, It has been modified to disable ExecNetwork level GetMetric API in this PR.

File: inference-engine/tests/functional/plugin/gpu/shared_tests_instances/behavior/plugin/core_integration.cpp [hunk 4]
```diff
@@ -291,6 +291,66 @@ INSTANTIATE_TEST_SUITE_P(
...
+    std::cout << "Memory Statistics after the 2nd and 3rd networks are loaded in parallel through threads: " << std::endl;
+    for (auto &&kv : t2) {
+        ASSERT_NE(kv.second, 0);
+        auto iter = t1.find(kv.first);
+        if (iter != t1.end()) {
+            ASSERT_EQ(kv.second, t1[kv.first] * 3);
+        }
+        std::cout << kv.first << ": " << kv.second << " bytes" << std::endl;
+    }
+
```
Comment (yeonbok) [inline inference-engine/tests/functional/plugin/gpu/shared_tests_instances/behavior/plugin/core_integration.cpp:345]: how about to check whether the result at this point is to be x3 of the first result?

Comment (andrew-k-park) [reply inference-engine/tests/functional/plugin/gpu/shared_tests_instances/behavior/plugin/core_integration.cpp:330]: As discussed, the code that prints memory statistics result has been removed from  the rest of test cases except for the default test case.

File: inference-engine/src/cldnn_engine/cldnn_engine.cpp (hunk 2 above)
Comment (vladimir-paramuzov) [reply inference-engine/src/cldnn_engine/cldnn_engine.cpp:281]: I meant GetMetric() call for multiple networks which are loaded sequentially

**Comment by andrew-k-park** [comment]
@vladimir-paramuzov All comments are applied and CI checks are passed. Could you review this PR again?

**Review by vladimir-paramuzov — APPROVED** [review]

**Pull request merged**